### PR TITLE
feat(puffin): apply range reader

### LINKED ISSUE #4594: Fulltext stager's cache behavior is in conflict with OpenDAL's cache layer
### What type of bug is this?

Other

### What subsystems are affected?

Datanode

### Minimal reproduce step

- create a table with fulltext column
- write data
- query data

### What did you expect to see?

no duplicate ranges in cached files

### What did you see instead?

too many duplications ranges like `0 ~ 20480`, `4 ~ 20480`

2G data files will consume ~60G disk cache

### What operating system did you use?

ArchLinux AMD64

### What version of GreptimeDB did you use?

nightly

### Relevant log output and stack trace

_No response_

### LINKED ISSUE #4717: [Tracking Issue] `RangeReader` as input source of Index
### What type of enhancement is this?

Performance

### What does the enhancement do?

Ref: https://github.com/GreptimeTeam/greptimedb/issues/4594

- [x] `RangeReader` trait
- [x] Adapter from `AsyncRead + AsyncSeek` to `RangeReader`
- [x] Top-down conversion of Input Source of `Index` from `AsyncRead + AsyncSeek` to `RangeReader`
- [x] Implement `RangeReader` for `ObjectStore`
- [x] Instrument `RangeReader`
- [x] Remove the adapter

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

close #4717
close #4594

## What's changed and what's your intention?

* Replace all `AsyncRead + AsyncSeek` Source with `RangeReader`
* Impls `RangeReader` for `ObjectStore` and `File`
* Adapts `AsyncRead` for `RangeReader` to make `handle_decompress` work

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Review by WenyXu — APPROVED** [review]
LGTM

File: src/puffin/src/partial_reader/async.rs [hunk 1]
```diff
@@ -13,184 +13,122 @@
...
+            .iter()
+            .map(|range| {
+                let start = self.offset + range.start;
+                let end = (self.offset + range.end).min(self.offset + self.size);
+                start..end
+            })
+            .collect();
+
+        for range in &absolute_ranges {
+            if range.start >= self.offset + self.size {
```
Comment (fengjiachun) [inline src/puffin/src/partial_reader/async.rs:79]: Question: why not check this in line 72?

**Review by fengjiachun — APPROVED** [review]
Rest LGTM

File: src/common/base/src/range_read.rs [hunk 2]
```diff
@@ -80,26 +88,162 @@ impl<R: RangeReader + Send + Unpin> RangeReader for &mut R {
...
 
-    async fn read(&mut self, range: Range<u64>) -> io::Result<Bytes> {
+    async fn read(&mut self, mut range: Range<u64>) -> io::Result<Bytes> {
+        range.end = range.end.min(self.len() as u64);
+
         let mut buf = vec![0; (range.end - range.start) as usize];
-        self.0.seek(io::SeekFrom::Start(range.start)).await?;
-        self.0.read_exact(&mut buf).await?;
+        buf.copy_from_slice(&self[range.start as usize..range.end as usize]);
+        Ok(Bytes::from(buf))
```
Comment (evenyag) [inline src/common/base/src/range_read.rs:196]: Maybe use `to_vec()`
```rust
let bytes = Bytes::copy_from_slice(&self[range.start as usize..range.end as usize]);
Ok(bytes)
```

File: src/common/base/src/range_read.rs [hunk 3]
```diff
@@ -80,26 +88,162 @@ impl<R: RangeReader + Send + Unpin> RangeReader for &mut R {
...
+            return Poll::Ready(Ok(to_read));
+        }
+
+        if this.read_fut.is_none() {
+            let range = *this.position..*this.content_length;
+            let inner = this.inner.clone();
+            let fut = async move {
+                let mut inner = inner.lock().await;
+                inner.read(range).await
+            };
```
Comment (evenyag) [inline src/common/base/src/range_read.rs:152]: If the content_length is too large, reading the whole range may cause oom. Maybe we should limit the max size to read.

File: src/common/base/Cargo.toml [hunk 4]
```diff
@@ -16,8 +16,10 @@ common-error.workspace = true
 common-macro.workspace = true
 futures.workspace = true
 paste = "1.0"
+pin-project.workspace = true
 serde = { version = "1.0", features = ["derive"] }
 snafu.workspace = true
+tokio.workspace = true
```
Comment (evenyag) [inline src/common/base/Cargo.toml:22]: Now we have two more dependencies (also the tokio) of the common-base 😢 

Maybe we should move the range_read mod or its implementations to object-store crate......

File: src/common/base/src/range_read.rs
Comment (evenyag) [inline src/common/base/src/range_read.rs:1]: Do we need to add some tests for the implementations of the RangeReader?

File: src/puffin/src/partial_reader/async.rs [hunk 5]
```diff
@@ -13,184 +13,122 @@
...
+    ) -> io::Result<()> {
+        let absolute_range_start = self.offset + range.start;
+        if absolute_range_start >= self.offset + self.size {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Start of range is out of bounds",
+            ));
         }
+        let absolute_range_end = (self.offset + range.end).min(self.offset + self.size);
+        let absolute_range = absolute_range_start..absolute_range_end;
```
Comment (evenyag) [inline src/puffin/src/partial_reader/async.rs:61]: Maybe we can implement a method to get absolute_range from range.

File: src/puffin/src/partial_reader/async.rs [hunk 6]
```diff
@@ -13,184 +13,122 @@
...
-        self.position_in_portion = Some(self.position() + read_bytes as u64);
+    async fn read_vec(&mut self, ranges: &[Range<u64>]) -> io::Result<Vec<Bytes>> {
+        let absolute_ranges: Vec<Range<u64>> = ranges
+            .iter()
+            .map(|range| {
+                let start = self.offset + range.start;
+                let end = (self.offset + range.end).min(self.offset + self.size);
+                start..end
+            })
+            .collect();
```
Comment (evenyag) [inline src/puffin/src/partial_reader/async.rs:76]: Can we collect it into `Resut<Vec<>>` so we can check it in map?

File: src/puffin/src/partial_reader/async.rs (hunk 1 above)
Comment (zhongzc) [reply src/puffin/src/partial_reader/async.rs:79]: udpated.

File: src/common/base/src/range_read.rs (hunk 2 above)
Comment (zhongzc) [reply src/common/base/src/range_read.rs]: updated

File: src/common/base/src/range_read.rs (hunk 3 above)
Comment (zhongzc) [reply src/common/base/src/range_read.rs]: added

**Pull request merged**